src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +563,544 @@ void ChunkScrub::chunk_scrub_common()
   cout << "--done--" << std::endl;
 }
 
+#define DEBUG_OUT(x) if(debug==1){std::cout<<x;}else{}
+
+class SampleDedup : public CrawlerThread
+{
+public:
+  enum class crawl_mode_t {
+    DEEP,
+    SHALLOW,
+  };
+
+  SampleDedup(
+    IoCtx& io_ctx,
+    IoCtx& chunk_io_ctx,
+    int n,
+    int m,
+    ObjectCursor& begin,
+    ObjectCursor end,
+    int32_t report_period,
+    uint64_t num_objects,
+    uint32_t sampling_ratio,
+    uint32_t object_dedup_threshold,
+    uint32_t chunk_dedup_threshold,
+    size_t chunk_size,
+    std::string& fp_algo,
+    crawl_mode_t mode) :
+    CrawlerThread(io_ctx, n, m, begin, end, report_period, num_objects),
+    chunk_io_ctx(chunk_io_ctx),
+    mode(mode),
+    sampling_ratio(sampling_ratio),
+    chunk_dedup_threshold(chunk_dedup_threshold),
+    object_dedup_threshold(object_dedup_threshold),
+    chunk_size(chunk_size),
+    fp_type(get_fp_type(fp_algo)) { }
+  static void clear_fingerprint_store() {
+    {
+      std::unique_lock lock(fingerprint_lock);
+      instant_fingerprint_store.clear();
+    }
+    {
+      std::unique_lock lock(flushed_lock);
+      flushed_objects.clear();
+    }
+  }
+
+  ~SampleDedup() { };
+
+protected:
+  void* entry() override {
+    crawl();
+    return NULL;
+  }
+
+private:
+  struct chunk_t {
+    string oid = "";
+    size_t start = 0;
+    size_t size = 0;
+    string fingerprint = "";
+    bufferlist data;
+  };
+
+  void crawl();
+  void prepare_rados();
+  std::tuple<ObjectCursor, ObjectCursor> get_shard_boundary();
+  std::tuple<std::vector<ObjectItem>, ObjectCursor> get_objects(
+    ObjectCursor current,
+    ObjectCursor end,
+    size_t max_object_count);
+  std::set<size_t> sample_object(size_t count);
+  void try_dedup_and_accumulate_result(ObjectItem& object);
+  bool ok_to_dedup_all();
+  void flush_duplicable_object(ObjectItem& object);
+  AioCompletion* set_chunk(chunk_t& chunk);
+  void mark_non_dedup(ObjectCursor start, ObjectCursor end);
+  bufferlist read_object(ObjectItem& object);
+  std::vector<std::tuple<bufferlist, pair<uint64_t, uint64_t>>> do_cdc(
+    ObjectItem& object,
+    bufferlist& data);
+  std::string generate_fingerprint(bufferlist chunk_data);
+  bool check_duplicated(std::string& fingerprint);
+  void add_duplication(chunk_t& chunk);
+  void add_fingerprint(chunk_t& chunk);
+  bool check_whole_object_dedupable(size_t dedup_size, size_t total_size);
+  bool is_dirty(ObjectItem& object);
+  AioCompletion* do_evict(string oid);
+
+  enum class fp_type_t {
+    SHA1,
+    SHA256,
+    SHA512,
+    UNKNOWN,
+  };
+
+  fp_type_t get_fp_type(string fp_algo);
+
+  Rados rados;
+  IoCtx chunk_io_ctx;
+  crawl_mode_t mode;
+  uint32_t sampling_ratio;
+  std::list<chunk_t> duplicable_chunks;
+  size_t total_duplicated_size = 0;
+  size_t total_object_size = 0;
+  size_t chunk_dedup_threshold;
+  size_t object_dedup_threshold;
+
+  struct fp_store_entry_t {
```

Comment:
hm.. how about adding something like dedup_info_store calss  to update/read information about collected fingerprints and dedupable object ? This class might include fp_store_entry_t.